insights/internal/uploader/upload.go

```diff
@@ -94,10 +101,15 @@ func (um Uploader) Upload(source string, force bool) error {
 	mu := &sync.Mutex{}
 	var uploadError error
 	var wg sync.WaitGroup
+	sem := make(chan struct{}, um.maxConcurrentUploadsPerSource)
+
 	for _, r := range reports {
 		wg.Add(1)
+		sem <- struct{}{} // Acquire a semaphore slot.
```

Comment:
To prevent allocating too many resources, I think we can:

sem := make(chan struct{}, min(len(reports), um.maxConcurrentUploadsPerSource))

Or similar. That may affect your test cases.